### PR TITLE
feat: Error revamp

### PR DESCRIPTION
**Pull request by jbelkins**
## Issue \#
#692

## Description of changes
This PR is the counterpart of https://github.com/awslabs/smithy-swift/pull/559 in `smithy-swift`.

Major changes:
- Code generation & codegen tests have been changed to match changes in the `smithy-swift` PR above.
- `AWSServiceError` extends `ServiceError` (from `smithy-swift`) with AWS-specific error parameters.  S3 & other customizations extend from `AWSServiceError`.
- Waiter error matching tests now use the `ServiceError` protocol to match errors.
- Adds a few S3 integration tests that cause, catch & examine errors.

## New/existing dependencies impact assessment, if applicable
No new dependencies were added to this change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jbelkins** [comment]
```mermaid
---
title: Error Protocols of the AWS SDK for Swift
---
classDiagram

class `Swift.Error` {
    <<protocol>>
    var localizedDescription String
}

class `ClientRuntime.ServiceError` {
    <<protocol>>
    var _errorType: String?
    var _message: String?
}
`Swift.Error` <|.. `ClientRuntime.ServiceError`
`ClientRuntime.HTTPResponseError` <|.. `ClientRuntime.ServiceError` : when wire protocol is HTTP
`ClientRuntime.RetryableError` <|.. `ClientRuntime.ServiceError`

class `ClientRuntime.HTTPResponseError` {
    <<protocol>>
    var _httpResponse: ClientRuntime.HttpResponse
}
`Swift.Error` <|.. `ClientRuntime.HTTPResponseError`

class `ClientRuntime.RetryableError` {
    <<protocol>>
    var _isRetryable: Bool
    var _isThrottlingError: Bool
}
`Swift.Error` <|.. `ClientRuntime.RetryableError`

class `ClientRuntime.SDKInternalError` {
    <<struct>>
    No properties
}
`Swift.Error` <|.. `ClientRuntime.SDKInternalError`
`ClientRuntime.RetryableError` <|.. `ClientRuntime.SDKInternalError` : not retryable, not throttling

class `ClientRuntime.ModeledError` {
    <<protocol>>
    var _isModeledRetryable: Bool
    var _isModeledThrottling: Bool
}
`ClientRuntime.ServiceError` <|.. `ClientRuntime.ModeledError`

class `AWSS3.InvalidObjectState` {
    <<struct; example modeled error>>
    custom modeled properties
}
`ClientRuntime.ModeledError` <|.. `AWSS3.InvalidObjectState`

class `ClientRuntime.UnknownServiceError` {
    <<struct>>
    no properties
}
`ClientRuntime.ServiceError` <|.. `ClientRuntime.UnknownServiceError`
`ClientRuntime.RetryableError` <|.. `ClientRuntime.UnknownServiceError`

```

File: Tests/Core/AWSClientRuntimeTests/Protocols/RestJSON/RestJSONErrorTests.swift [hunk 1]
```diff
@@ -113,29 +107,19 @@ extension ComplexError {
...
 
-public enum GreetingWithErrorsError {
-    case complexError(ComplexError)
-    case unknown(UnknownAWSHttpServiceError)
-
-    public init(errorType: String?, httpResponse: HttpResponse, decoder: ResponseDecoder? = nil, message: String? = nil, requestID: String? = nil) throws {
-        switch errorType {
-        case "ComplexError" : self = .complexError(try ComplexError(httpResponse: httpResponse, decoder: decoder, message: message, requestID: requestID))
-        default : self = .unknown(UnknownAWSHttpServiceError(httpResponse: httpResponse, message: message))
+public enum GreetingWithErrorsError: HttpResponseErrorBinding {
```
Comment (jbelkins) [inline Tests/Core/AWSClientRuntimeTests/Protocols/RestJSON/RestJSONErrorTests.swift:116]: Changing test errors to the new format

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/awsjson/AWSJsonHttpResponseBindingErrorGenerator.kt [hunk 2]
```diff
@@ -9,38 +9,53 @@ import software.amazon.smithy.aws.swift.codegen.AWSClientRuntimeTypes
...
 import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
 import software.amazon.smithy.swift.codegen.integration.httpResponse.HttpResponseBindingErrorGeneratable
 import software.amazon.smithy.swift.codegen.model.toUpperCamelCase
+import software.amazon.smithy.swift.codegen.utils.errorShapeName
 
-/*
- * TODO: The code below is not valid for AWS Json 1.0/1.1.
- *       This code was lifted from restJson to make tests compile and run for Json 1.0/1.1
- */
 class AWSJsonHttpResponseBindingErrorGenerator : HttpResponseBindingErrorGeneratable {
```
Comment (jbelkins) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/awsjson/AWSJsonHttpResponseBindingErrorGenerator.kt:21]: This is the AWS JSON error generator.  You will see similar code below for other AWS protocols.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/route53/Route53InvalidBatchErrorIntegration.kt [hunk 3]
```diff
@@ -27,14 +27,11 @@ class Route53InvalidBatchErrorIntegration : SwiftIntegration {
     private val httpResponseBindingErrorGenerator = SectionWriter { writer, previousCode ->
         val operationErrorName = writer.getContext("operationErrorName") as String
         if (operationErrorName == "ChangeResourceRecordSetsOutputError") {
```
Comment (jbelkins) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/route53/Route53InvalidBatchErrorIntegration.kt:29]: I discovered that this Route 53 customization isn't currently working.

Wrote a ticket to get it under test & verified working:
https://github.com/awslabs/aws-sdk-swift/issues/1027

File: codegen/smithy-aws-swift-codegen/src/test/resources/software.amazon.smithy.aws.swift.codegen/route53-invalidbatch.smithy [hunk 4]
```diff
@@ -4,7 +4,7 @@ namespace com.amazonaws.route53
 use aws.api#service
 use aws.protocols#restXml
 
-@service(sdkId: "Route53")
+@service(sdkId: "Route 53")
```
Comment (jbelkins) [inline codegen/smithy-aws-swift-codegen/src/test/resources/software.amazon.smithy.aws.swift.codegen/route53-invalidbatch.smithy:7]: `Route 53` (with space) is the actual sdkId for Route 53.

Comment (jbelkins) [reply codegen/smithy-aws-swift-codegen/src/test/resources/software.amazon.smithy.aws.swift.codegen/route53-invalidbatch.smithy:7]: This whole Route 53 customization is not tested, that will happen with this ticket:
https://github.com/awslabs/aws-sdk-swift/issues/1027

**Review by waahm7 — APPROVED** [review]

File: IntegrationTests/Services/AWSS3IntegrationTests/S3ErrorTests.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+class S3ErrorTests: S3XCTestCase {
+
+    func test_noSuchKey_throwsNoSuchKeyWhenUnknownKeyIsUsed() async throws {
+        do {
+            let input = GetObjectInput(bucket: bucketName, key: UUID().uuidString)
+            _ = try await client.getObject(input: input)
+            XCTFail("Request should not have succeeded")
+        } catch let error as NoSuchKey {
+            XCTAssertEqual(error.httpResponse.statusCode, .notFound)
+            XCTAssertEqual(error.message, "The specified key does not exist.")
```
[the remark this answers is not in the document]
Comment (jbelkins) [reply IntegrationTests/Services/AWSS3IntegrationTests/S3ErrorTests.swift:22]: I thought about that too.  It is very possible that the message could change & break the test.  But I want to ensure that the integration tests verify that the actual message can be read, so I'm willing to risk the breakage.

**Pull request merged**